widget/embedded/src/store/utils/data.ts

```diff
@@ -56,47 +98,32 @@ export function matchTokensFromConfigWithMeta(params: {
    * From this point onward,
    * the function handles the other type of widget configuration.
    */
-  if (!config.blockchains || config.blockchains.length === 0) {
+  if (config.blockchains?.length) {
+    if (configTokens) {
+      handleConfigTokensObject(configTokens, config.blockchains);
+      return Object.values(result);
+    }
+  } else {
     /**
-     * If "config.blockchains" does not exist,
-     * there are no limitations on the tokens, and we should include all tokens from the meta response.
+     * If config.blockchains does not exist,
+     * We handle the tokens object as in the previous branch.
+     * Then we include all tokens from every blockchain in the meta data that are not specified in the tokens from the config.
      */
-    return Array.from(meta.tokensMapByTokenHash.values());
-  }
 
-  //We iterate over each blockchain and retrieve the related tokens for each one from the configuration.
-  config.blockchains.forEach((blockchain) => {
-    const targetTokensForBlockchain = configTokens?.[blockchain];
-    /**
-     * If no tokens exist for a blockchain,
-     * we include all tokens for that blockchain from the meta response.
-     */
-    if (
-      !targetTokensForBlockchain &&
-      meta.tokensMapByBlockchainName?.[blockchain]
-    ) {
-      addTokens(meta.tokensMapByBlockchainName[blockchain]);
-      return;
-    }
+    if (configTokens) {
+      const blockchains = new Set(Object.keys(configTokens));
 
-    if (targetTokensForBlockchain) {
-      if (targetTokensForBlockchain.isExcluded) {
-        /**
-         * If tokens are excluded,
-         * we first include all tokens from the meta for that blockchain,
-         * then remove the excluded tokens from the result.
-         */
-        addTokens(meta.tokensMapByBlockchainName[blockchain]);
-        targetTokensForBlockchain.tokens.forEach((token) => {
-          const tokenHash = createTokenHash(token);
-          delete result[tokenHash];
-        });
-      } else {
-        //We retrieve the corresponding token from the meta and include it in the result.
-        addTokens(targetTokensForBlockchain.tokens);
-      }
+      handleConfigTokensObject(configTokens, Array.from(blockchains));
```

Comment:
This is really good to reuse the logic here, but because both `handleConfigTokensObject` and `addTokens` are implicitly update environment (`result` variable), it made the code less readable and harder to follow.

I suggest two things here: 
1. if you want to keep the function (`handleConfigTokensObject`), rename to a more descriptive name and also make the function pure by returning `result`  instead of mutating the parent scope. I mean `function handleConfigTokensObject(...): Record<TokenHash, Token>` then update the variable like this `result = handleConfigTokensObject(...)`.
2. Your code can be restructured to not need to make a separate function for reusing the code (`handleConfigTokensObject`). 
It can be something like this:

```js

let configBlockchains;
if (config.blockchains?.length) {
configBlockchains = config.blockchains;
}else{
configBlockchains = new Set(Object.keys(configTokens));

 Object.keys(meta.tokensMapByBlockchainName).forEach((blockchain) => {
        if (!blockchains.has(blockchain)) {
          addTokens(meta.tokensMapByBlockchainName[blockchain]);
        }
      });
}

// Rest of code that has been moved to `handleConfigTokensObject`

```